### DISCUSSION TITLE
Why Python and not R

### DISCUSSION THREAD

**Discussion by gitboy16** [IyarLin/R-advantages-over-python#1] [General]
Please don't get me wrong. I love R and at my company we both use R and Python but R users in our corporate environment are slowly moving to Python. Why? The reason is simple. Licensing. The following article summarizes it:
https://www.r-bloggers.com/2019/01/how-gpl-makes-me-leave-r-for-python/
A couple of other reasons are: Python community is much larger than R and hence can respond more quickly to users issue. A lot of the most used R package have unadressed opened issue...very often ignored. It also exists on the Python side but it is easier to find someone to help you because of the community. Also when you need to hire someone, finding someone who is good at python is much easier than finding someone good at R.
Finally if you are looking for a job, you can't go wrong with Python. R is a niche programming language. I also see that Java, C++,C# are very much in demand. Finally watch out for Julia and Rust. These 2, although still very new, are coming with very strong arguments and liberal licenses.

**Comment by IyarLin** [comment]
Thanks for the feedback! I fully agree with some of the challenges you mentioned such as difficulty of hiring proficient R users compared with python users. I think that has a lot to do with the snowball dynamic I mentioned. It becomes a chicken and egg situation.

Re the liscensing issue: could you please elaborate on that? I know there are a lot of companies using R commercially so I wonder what are the specific use cases/products where the liscense becomes an issue.

Also your mention of Rust is interesting. I heard about Julia for DS (though it's promise doesn't appeal to me as I rarely encounter computation time bottlenecks outside big data) but never of Rust in the context of DS.

**Reply by gitboy16** [reply]
I have attached a link about licensing in my previous post. Everything is explained there.
Quoting - "GPL is not a permissive license. It is categorized as “strongly protective”.
In layman terms, if you build your work on a GPL program it may force you to license your product with a GPL license, too. In other words – it restrains you from keeping your code proprietary." - The reality is that GPL is confusing for a lot of developpers and companies don't want the headache to deal with licensing stuff so they simply turn to something that they know, can change, relicense and sell as they wish. Another thing as well, is that Python is very general, you can build an .exe and ship it or you can build an android app,Third party libraries are very often build in C/C++, Java and Python not in R. Python is very often installed by default on major linux distributions as it is great for scripting. Python has spread across various communities and try to be as inclusive and open as possible. Some who choose Python over R can't go wrong. I would have loved to see that with R...
R is a great language and I use it when I can. (Something however I hate is that long vector support is really bad in R, when are we going to see a data.frame with long vector support? I mean come on, we are in the 21st century! Looks like R is still stuck in the previous century with its int32)

**Reply by IyarLin** [reply]
Are you sure long vectors are not supported in R? I know there's the int64 package and also I found this:

https://stat.ethz.ch/R-manual/R-devel/library/base/html/LongVectors.html

As for being able to write android apps in python or general software development I wonder how often is that relevant to the work of a DS?

The liscensing issue does sound like a serious barrier to companies selling software, I agree on that.

**Reply by gitboy16** [reply]
The int64 package is not on cran anymore. If you want to use something close to integer64 you can use bit64. Long vector is supported in some R base function basically by turning int into double...this is limited to some specific functions. `data.table` has better support than base R because it uses bit64 but still you can't really create a data.frame with more than 2^31-1 row. See https://github.com/Rdatatable/data.table/issues/3957

What I meant by being able to write android apps is that python is more a general purpose programming language which enable you to much more than DS. Python has managed to build brigdes between programmer communities and be inclusive of all type of programmers, analyst, quants, engineer, scientist..etc they all can talk the same language and this makes working together much easier....anyway I think people must simply use what they want, like and feel comfortable with...

**Comment by gitboy16** [comment]
https://m.youtube.com/watch?v=6qCH7Y2rc_w
